### PR TITLE
feat: add support for CoRE RD discovery

### PR DESCRIPTION
**Pull request by JKRhb**
Besides some cleanup, this PR introduces the possibility to discover CoRE RDs themselves (or rather: their resource lookup interfaces) during the discovery process with a new discovery method. This enables users to discover TDs in a three-step process:

1. Obtain one or more links pointing to lookup interfaces
2. Obtain links pointing to TDs from the lookup interface(s)
3. Retrieve the TDs from the resulting links

I haven't been able to properly test the new implementation with actual devices and RDs, however, I ran sample requests using hard-coded values in the discovery methods while having Wireshark open and the correct requests are being sent.

(This is probably the last "major" feature relevant for our project, by the way. After this PR, I will probably mostly make some minor changes/additions with regard to the TD data models.)

### CODE REVIEW

**Comment by JKRhb** [comment]
Applied a couple of additional refactorings, getting rid of some redundancy and fixing a bug. The PR should now actually be finished.

**Comment by JKRhb** [comment]
Noticed and fixed a couple of additional small issues.

**Comment by JKRhb** [comment]
Managed to test out the changes with an actual device now and discovered a few issues, which I have fixed in the latest updated commits.

File: lib/src/core/thing_discovery.dart [hunk 1]
```diff
@@ -86,42 +102,90 @@ class ThingDiscovery extends Stream<ThingDescription>
...
-          host: uri.host,
-          port: uri.port,
-        );
-      }
-      if (discoveredUris.contains(parsedUri)) {
-        continue;
+      for (final parsedUri in parsedUris) {
+        if (discoveredUris.contains(parsedUri)) {
+          continue;
+        }
```
Comment (falko17) [inline lib/src/core/thing_discovery.dart:184]: Is this check really necessary? `discoveredUris` is already a `Set`, so adding an element which is already present shouldn't have any effect.

Comment (JKRhb) [reply lib/src/core/thing_discovery.dart]: Hmm, very good point regarding the `Set`. I refactored the code in 83d6125a185e530334d28b7bf3e4944d16331934, taking advantage of the boolean return value of the `add` method and making it a bit more concise. Hope that this is an improvement :)

Comment (falko17) [reply lib/src/core/thing_discovery.dart]: Yes, that seems like a good solution 👍🏻

**Pull request merged**